{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "from arcgis.gis import GIS\n",
    "#from arcgis.features import GeoAccessor, GeoSeriesAccessor\n",
    "from arcgis.features import FeatureLayerCollection\n",
    "import pandas as pd\n",
    "import arcpy, os, time, uuid\n",
    "from zipfile import ZipFile\n",
    "import arcgis.features\n",
    "import sys, os, shutil\n",
    "\n",
    "# Start Timer\n",
    "startTime = time.time()\n",
    "\n",
    "gis =GIS('PRO')\n",
    "\n",
    "existe=True\n",
    "lst=pd.read_html('https://www.estar.cl/fire/planilla.php')\n",
    "df=pd.DataFrame(lst[0],columns=['TEMPORADA',\t'INICIO',\t'SEMANA INICIO',\t'SEMANA KPI',\t'ANO INICIO',\t'MES INICIO',\t'DIA INICIO',\t'HORA INICIO',\t'COD_SEMANA',\t'DIA_SEMANA',\t'RANGO_PREV',\t'REGION',\t'PROVINCIA',\t'COMUNA',\t'EMPRESA',\t'PROPIETARIO',\t'TIPO',\t'AREA',\t'FICHA',\t'COD_PREDIO',\t'PREDIO',\t'FOCOS',\t'MULTIFOCO',\t'DANO PLANTAC',\t'CAUSA',\t'CODIGO_CAUSA',\t'CONFLICTO',\t'GRUPO CAUSAL',\t'ZONA META',\t'COD META',\t'COMBUSTIBLE INICIAL',\t'COMBUSTIBLE FINAL',\t'DESPACHADOR',\t'DETECTOR',\t'AVISO',\t'COMANDANTE',\t'ANCHO LINEA',\t'LARGO LINEA',\t'CLASE',\t'GASTO COMBATE UF',\t'FECHA CONTROL',\t'HORA CONTROL',\t'LONGITUD',\t'LATITUD',\t'ESTADO',\t'SUP INICIAL',\t'PIRA',\t'EUCA',\t'OTRAS',\t'TOTAL',\t'SUM DANO',\t'USD DANO',\t'ESTADO DANO',\t'DETECCION',\t'DESPACHO',\t'ARRIBO',\t'CONTENCION',\t'CONTROL',\t'EXTINCION',\t'MIN DESPACHO',\t'PROM SALIDA',\t'MIN ARRIBO',\t'PROM INICIO COMBATE',\t'MIN CONTENCION',\t'MIN CONTROL',\t'MIN EXTINCION',\t'MIN HASTA CONTENCION',\t'MIN TOTAL',\t'RECURSOS DESPACHO',\t'RECURSOS COMBATE',\t'RECURSOS CONTROL',\t'ESTIMADO CONTROL',\t'AREA CONTROL'])\n",
    "# Convert the values to real coordinate values\n",
    "df['LONGITUD'] = df['LONGITUD'].astype(str).str[:3] + '.' + df['LONGITUD'].astype(str).str[3:]\n",
    "df['LATITUD'] = df['LATITUD'].astype(str).str[:3] + '.' + df['LATITUD'].astype(str).str[3:]\n",
    "\n",
    "\n",
    "# Define a lambda function to apply the logic\n",
    "def get_part_of_day(hour):\n",
    "    if hour < 7:\n",
    "        return \"Noche\"\n",
    "    elif hour < 12:\n",
    "        return \"Mañana\"\n",
    "    elif hour < 21:\n",
    "        return \"Tarde\"\n",
    "    else:\n",
    "        return \"Noche\"\n",
    "\n",
    "# Apply the lambda function to create the new column \"Parte del día\"\n",
    "df[\"Partedia\"] = df[\"HORA INICIO\"].apply(lambda x: get_part_of_day(x))\n",
    "\n",
    "import pandas as pd\n",
    "from datetime import datetime\n",
    "\n",
    "df['DateTime'] = pd.to_datetime(df['INICIO'])\n",
    "\n",
    "# Extract specific components\n",
    "df['Date'] = df['DateTime'].dt.date\n",
    "\n",
    "\n",
    "# Add a new column 'DayOfWeek' with the day of the week (Monday = 1 to Sunday = 7)\n",
    "df['DayOfWeek'] = df['Date'].apply(lambda x: x.weekday() + 1)\n",
    "df['ID'] = df.reset_index().index\n",
    "\n",
    "df_columns=['ID','TEMPORADA', 'INICIO', 'SEMANA KPI'\n",
    "            , 'EMPRESA', 'AREA', 'FICHA'\n",
    "            , 'COD_PREDIO', 'PREDIO', 'FOCOS', 'MULTIFOCO'\n",
    "            , 'CAUSA', 'ESTADO', 'PIRA'\n",
    "            , 'EUCA', 'OTRAS',\"HORA INICIO\", 'Partedia', 'DayOfWeek'\n",
    "            , 'LONGITUD', 'LATITUD']\n",
    "df_selected=df.loc[:,df_columns]\n",
    "df_rename={'TEMPORADA': 'Temporada'\n",
    "           , 'INICIO':'Fecha'\n",
    "           , 'SEMANA KPI':'Semana'\n",
    "           , 'EMPRESA':'Zona'\n",
    "           , 'AREA':'Paisaje'\n",
    "           , 'FICHA':'Ficha'\n",
    "           , 'COD_PREDIO':'idpredio'\n",
    "           , 'PREDIO':'Predio'\n",
    "           , 'FOCOS':'Focos'\n",
    "           , 'MULTIFOCO':'Multifocos'\n",
    "           , 'CAUSA':'Causa'\n",
    "           , 'ESTADO':'Estado'\n",
    "           , 'PIRA':'Pira ha'\n",
    "           , 'EUCA':'Euca ha'\n",
    "           , 'OTRAS':'Otras ha'\n",
    "           ,\"HORA INICIO\":'Hora inicio'\n",
    "           , 'Partedia':'Parde_del_dia'\n",
    "            , 'DayOfWeek':'Dia'\n",
    "            ,'LONGITUD':'longitude'\n",
    "            ,'LATITUD':'latitude'}\n",
    "\n",
    "df_selected=df_selected.rename(columns=df_rename)\n",
    "\n",
    "df3=pd.read_csv('data/lastupdate.csv')\n",
    "new_records = df_selected[~df_selected['Ficha'].isin(df3['Ficha'])]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "sdf=pd.DataFrame.spatial.from_xy(df_selected, x_column='longitude', y_column='latitude')\n",
    "temp_gdb = 'C:/temp/temp_wildfire.gdb'  # Change this path to your desired temporary location"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "if os.path.exists(temp_gdb):\n",
    "    shutil.rmtree(temp_gdb)\n",
    "else:\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "if len(new_records)>0:\n",
    "    arcpy.management.CreateFileGDB(os.path.dirname(temp_gdb), os.path.basename(temp_gdb))\n",
    "    sdf.spatial.to_featureclass(location=temp_gdb + '//temp_file', overwrite=True)\n",
    "else:\n",
    "    sys.exit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "33"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(new_records)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Connecting to AGOL\n",
      "Creating temporary File Geodatabase\n",
      "Exporting temp_file to temp FGD\n",
      "Zipping temp FGD\n",
      "Uploading File Geodatabase\n",
      "Truncating Feature Service\n",
      "Get feature service fields\n",
      "Get feature class/table fields\n",
      "Adding additional fields\n",
      "\tAdding field pira_ha\n"
     ]
    },
    {
     "ename": "Exception",
     "evalue": "ERROR: Field name collision found: pira_ha.\n(Error Code: 500)",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mException\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-23-970694134dea>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m    174\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    175\u001b[0m             \u001b[0mupdate_dict\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;33m{\u001b[0m\u001b[1;34m\"fields\"\u001b[0m\u001b[1;33m:\u001b[0m \u001b[1;33m[\u001b[0m\u001b[0mnew_field\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m}\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 176\u001b[1;33m             \u001b[0mfLyr\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mmanager\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0madd_to_definition\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mupdate_dict\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    177\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    178\u001b[0m \u001b[1;31m# Append features from featureService class/hostedTable\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mc:\\Users\\raul.perez.astorga\\AppData\\Local\\ESRI\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages\\arcgis\\features\\managers.py\u001b[0m in \u001b[0;36madd_to_definition\u001b[1;34m(self, json_dict, future)\u001b[0m\n\u001b[0;32m   2499\u001b[0m         \u001b[0mu_url\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_url\u001b[0m \u001b[1;33m+\u001b[0m \u001b[1;34m\"/addToDefinition\"\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   2500\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 2501\u001b[1;33m         \u001b[0mres\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_con\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mpost\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mu_url\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mparams\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   2502\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mfuture\u001b[0m \u001b[1;32mand\u001b[0m \u001b[1;34m\"statusURL\"\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mres\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   2503\u001b[0m             \u001b[0mexecutor\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0m_cf\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mThreadPoolExecutor\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mc:\\Users\\raul.perez.astorga\\AppData\\Local\\ESRI\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages\\arcgis\\gis\\_impl\\_con\\_connection.py\u001b[0m in \u001b[0;36mpost\u001b[1;34m(self, path, params, files, **kwargs)\u001b[0m\n\u001b[0;32m   1077\u001b[0m             \u001b[0mfile_name\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mfile_name\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1078\u001b[0m             \u001b[0mtry_json\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mtry_json\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 1079\u001b[1;33m             \u001b[0mforce_bytes\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mpop\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"force_bytes\"\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;32mFalse\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   1080\u001b[0m         )\n\u001b[0;32m   1081\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mc:\\Users\\raul.perez.astorga\\AppData\\Local\\ESRI\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages\\arcgis\\gis\\_impl\\_con\\_connection.py\u001b[0m in \u001b[0;36m_handle_response\u001b[1;34m(self, resp, file_name, out_path, try_json, force_bytes, ignore_error_key)\u001b[0m\n\u001b[0;32m    623\u001b[0m                     \u001b[1;32mreturn\u001b[0m \u001b[0mdata\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    624\u001b[0m                 \u001b[0merrorcode\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mdata\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m\"error\"\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m\"code\"\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;32mif\u001b[0m \u001b[1;34m\"code\"\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mdata\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m\"error\"\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;32melse\u001b[0m \u001b[1;36m0\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 625\u001b[1;33m                 \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_handle_json_error\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m\"error\"\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0merrorcode\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    626\u001b[0m             \u001b[1;32mreturn\u001b[0m \u001b[0mdata\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    627\u001b[0m         \u001b[1;32melse\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mc:\\Users\\raul.perez.astorga\\AppData\\Local\\ESRI\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages\\arcgis\\gis\\_impl\\_con\\_connection.py\u001b[0m in \u001b[0;36m_handle_json_error\u001b[1;34m(self, error, errorcode)\u001b[0m\n\u001b[0;32m    646\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    647\u001b[0m         \u001b[0merrormessage\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0merrormessage\u001b[0m \u001b[1;33m+\u001b[0m \u001b[1;34m\"\\n(Error Code: \"\u001b[0m \u001b[1;33m+\u001b[0m \u001b[0mstr\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0merrorcode\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;33m+\u001b[0m \u001b[1;34m\")\"\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 648\u001b[1;33m         \u001b[1;32mraise\u001b[0m \u001b[0mException\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0merrormessage\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    649\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    650\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0mpost_multipart\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mpath\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mparams\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;32mNone\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mfiles\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;32mNone\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mException\u001b[0m: ERROR: Field name collision found: pira_ha.\n(Error Code: 500)"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "# # Overwrite Output\n",
    "\n",
    "# Overwrite Output\n",
    "arcpy.env.overwriteOutput = True\n",
    "\n",
    "fc = temp_gdb + '//temp_file'                                   # Path to Feature Class/Table\n",
    "fsItemId = \"d5b63197aea34575970a6b2b05cf1ab6\"                   # Feature Service Item ID to update\n",
    "featureService = True                                           # True if updating a Feature Service, False if updating a Hosted Table\n",
    "hostedTable = False                                             # True is updating a Hosted Table, False if updating a Feature Service\n",
    "layerIndex = 0                                                  # Layer Index\n",
    "disableSync = True                                              # True to disable sync, and then re-enable sync after append, False to not disable sync.  Set to True if sync is not enabled\n",
    "updateSchema = True                                             # True will remove/add fields from feature service keeping schema in-sync, False will not remove/add fields\n",
    "\n",
    "# Start Timer\n",
    "startTime = time.time()\n",
    "\n",
    "# Create GIS object\n",
    "print(\"Connecting to AGOL\")\n",
    "gis = GIS(\"PRO\")\n",
    "\n",
    "# Create UUID variable for GDB\n",
    "gdbId = str(uuid.uuid1())\n",
    "\n",
    "# Function to Zip FGD\n",
    "def zipDir(dirPath, zipPath):\n",
    "    '''Zip File Geodatabase'''\n",
    "    zipf = ZipFile(zipPath , mode='w')\n",
    "    gdb = os.path.basename(dirPath)\n",
    "    for root, _ , files in os.walk(dirPath):\n",
    "        for file in files:\n",
    "            if 'lock' not in file:\n",
    "               filePath = os.path.join(root, file)\n",
    "               zipf.write(filePath , os.path.join(gdb, file))\n",
    "    zipf.close()\n",
    "\n",
    "print(\"Creating temporary File Geodatabase\")\n",
    "gdb = arcpy.CreateFileGDB_management(arcpy.env.scratchFolder, gdbId)[0]\n",
    "\n",
    "# Export featureService classes to temporary File Geodatabase\n",
    "fcName = os.path.basename(fc)\n",
    "fcName = fcName.split('.')[-1]\n",
    "print(f\"Exporting {fcName} to temp FGD\")\n",
    "if featureService == True:\n",
    "    arcpy.conversion.FeatureClassToFeatureClass(fc, gdb, fcName)\n",
    "elif hostedTable == True:\n",
    "    arcpy.conversion.TableToTable(fc, gdb, fcName)\n",
    "\n",
    "# Zip temp FGD\n",
    "print(\"Zipping temp FGD\")\n",
    "zipDir(gdb, gdb + \".zip\")\n",
    "\n",
    "# Upload zipped File Geodatabase\n",
    "print(\"Uploading File Geodatabase\")\n",
    "fgd_properties={'title':gdbId, 'tags':'temp file geodatabase', 'type':'File Geodatabase'}\n",
    "fgd_item = gis.content.add(item_properties=fgd_properties, data=gdb + \".zip\")\n",
    "\n",
    "# Get featureService/hostedTable layer\n",
    "serviceLayer = gis.content.get(fsItemId)\n",
    "if featureService == True:\n",
    "    fLyr = serviceLayer.layers[layerIndex]\n",
    "elif hostedTable == True:\n",
    "    fLyr = serviceLayer.tables[layerIndex]\n",
    "\n",
    "# Truncate Feature Service\n",
    "# If views exist, or disableSync = False use delete_features.  OBJECTIDs will not reset\n",
    "flc = arcgis.features.FeatureLayerCollection(serviceLayer.url, gis)\n",
    "hasViews = False\n",
    "try:\n",
    "    if flc.properties.hasViews == True:\n",
    "        print(\"Feature Service has view(s)\")\n",
    "        hasViews = True\n",
    "except:\n",
    "    hasViews = False\n",
    "\n",
    "if hasViews == True or disableSync == False:\n",
    "    objectIdField = fLyr.manager.properties.objectIdField\n",
    "    # Get Min OBJECTID\n",
    "    minOID = fLyr.query(out_statistics=[{\"statisticType\": \"MIN\", \"onStatisticField\": objectIdField, \"outStatisticFieldName\": \"MINOID\"}])\n",
    "    minOBJECTID = minOID.features[0].attributes['MINOID']\n",
    "\n",
    "    # Get Max OBJECTID\n",
    "    maxOID = fLyr.query(out_statistics=[{\"statisticType\": \"MAX\", \"onStatisticField\": objectIdField, \"outStatisticFieldName\": \"MAXOID\"}])\n",
    "    maxOBJECTID = maxOID.features[0].attributes['MAXOID']\n",
    "\n",
    "    # If more than 2,000 features, delete in 2000 increments\n",
    "    if (maxOBJECTID - minOBJECTID) > 2000:\n",
    "        print(\"Deleting features\")\n",
    "        x = minOBJECTID\n",
    "        y = x + 1999\n",
    "        while x < maxOBJECTID:\n",
    "            query = f\"{objectIdField} >= {x} AND {objectIdField} <= {y}\"\n",
    "            fLyr.delete_features(where=query)\n",
    "            x += 2000\n",
    "            y += 2000\n",
    "    # Else if less than 2,000 features, delete all\n",
    "    else:\n",
    "        print(\"Deleting features\")\n",
    "        fLyr.delete_features(where=\"1=1\")\n",
    "\n",
    "# If no views and disableSync is True: disable Sync, truncate, and then re-enable Sync.  OBJECTIDs will reset\n",
    "elif hasViews == False and disableSync == True:\n",
    "    if flc.properties.syncEnabled == True:\n",
    "        print(\"Disabling Sync\")\n",
    "        properties = flc.properties.capabilities\n",
    "        updateDict = {\"capabilities\": \"Query\", \"syncEnabled\": False}\n",
    "        flc.manager.update_definition(updateDict)\n",
    "        print(\"Truncating Feature Service\")\n",
    "        fLyr.manager.truncate()\n",
    "        print(\"Enabling Sync\")\n",
    "        updateDict = {\"capabilities\": properties, \"syncEnabled\": True}\n",
    "        flc.manager.update_definition(updateDict)\n",
    "    else:\n",
    "        print(\"Truncating Feature Service\")\n",
    "        fLyr.manager.truncate()\n",
    "\n",
    "# Schema Sync\n",
    "if updateSchema == True:\n",
    "    # Get feature service fields\n",
    "    print(\"Get feature service fields\")\n",
    "    featureServiceFields = {}\n",
    "    for field in fLyr.manager.properties.fields:\n",
    "        if field.type != 'esriFieldTypeOID' and 'Shape_' not in field.name and 'GlobalID' not in field.name:\n",
    "            featureServiceFields[field.name] = field.type\n",
    "\n",
    "    # Get feature class/table fields\n",
    "    print(\"Get feature class/table fields\")\n",
    "    featureClassFields = {}\n",
    "    arcpy.env.workspace = gdb\n",
    "    if hostedTable == True:\n",
    "        for field in arcpy.ListFields(fc):\n",
    "            if field.type != 'OID' and field.type != 'Geometry' and 'GlobalID' not in field.name:\n",
    "                featureClassFields[field.name] = field.type\n",
    "    else:\n",
    "        for field in arcpy.ListFields(fc):\n",
    "            if field.type != 'OID' and field.type != 'Geometry' and 'Shape_' not in field.name and 'GlobalID' not in field.name:\n",
    "                featureClassFields[field.name] = field.type\n",
    "\n",
    "    minusSchemaDiff = set(featureServiceFields) - set(featureClassFields)\n",
    "    addSchemaDiff = set(featureClassFields) - set(featureServiceFields)\n",
    "\n",
    "    # Delete removed fields\n",
    "    if len(minusSchemaDiff) > 0:\n",
    "        print(\"Deleting removed fields\")\n",
    "        for key in minusSchemaDiff:\n",
    "            print(f\"\\tDeleting field {key}\")\n",
    "            remove_field = {\n",
    "                \"name\": key,\n",
    "                \"type\": featureServiceFields[key]\n",
    "            }\n",
    "            update_dict = {\"fields\": [remove_field]}\n",
    "            fLyr.manager.delete_from_definition(update_dict)\n",
    "\n",
    "    # Create additional fields\n",
    "    fieldTypeDict = {}\n",
    "    fieldTypeDict['Date'] = 'esriFieldTypeDate'\n",
    "    fieldTypeDict['Double'] = 'esriFieldTypeDouble'\n",
    "    fieldTypeDict['Integer'] = 'esriFieldTypeInteger'\n",
    "    fieldTypeDict['String'] = 'esriFieldTypeString'\n",
    "    if len(addSchemaDiff) > 0:\n",
    "        print(\"Adding additional fields\")\n",
    "        for key in addSchemaDiff:\n",
    "            print(f\"\\tAdding field {key}\")\n",
    "            if fieldTypeDict[featureClassFields[key]] == 'esriFieldTypeString':\n",
    "                new_field = {\n",
    "                    \"name\": key,\n",
    "                    \"type\": fieldTypeDict[featureClassFields[key]],\n",
    "                    \"length\": [field.length for field in arcpy.ListFields(fc, key)][0]\n",
    "                }\n",
    "            else:\n",
    "                new_field = {\n",
    "                    \"name\": key,\n",
    "                    \"type\": fieldTypeDict[featureClassFields[key]]\n",
    "                }\n",
    "\n",
    "            update_dict = {\"fields\": [new_field]}\n",
    "            fLyr.manager.add_to_definition(update_dict)\n",
    "\n",
    "# Append features from featureService class/hostedTable\n",
    "print(\"Appending features\")\n",
    "#fLyr.append(item_id=fgd_item.id, upload_format=\"filegdb\", upsert=False, field_mappings=[])\n",
    "arcpy.management.Append(temp_gdb + '//temp_file', \"https://araucaria.arauco.com/server/rest/services/Hosted/OcurrenciaIncendiosTActual/FeatureServer/0\", \"NO_TEST\")\n",
    "# Delete Uploaded File Geodatabase\n",
    "print(\"Deleting uploaded File Geodatabase\")\n",
    "fgd_item.delete()\n",
    "\n",
    "# Delete temporary File Geodatabase and zip file\n",
    "print(\"Deleting temporary FGD and zip file\")\n",
    "arcpy.Delete_management(gdb)\n",
    "arcpy.management.Delete(temp_gdb)\n",
    "os.remove(gdb + \".zip\")\n",
    "\n",
    "endTime = time.time()\n",
    "elapsedTime = round((endTime - startTime) / 60, 2)\n",
    "print(\"Script finished in {0} minutes\".format(elapsedTime))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'ee'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-14-d19a1019df43>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[1;32mimport\u001b[0m \u001b[0mee\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      2\u001b[0m \u001b[1;32mimport\u001b[0m \u001b[0mgeemap\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      3\u001b[0m \u001b[0mee\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mAuthenticate\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      4\u001b[0m \u001b[1;31m#ee.Initialize()\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      5\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mModuleNotFoundError\u001b[0m: No module named 'ee'"
     ]
    }
   ],
   "source": [
    "import ee\n",
    "import geemap\n",
    "ee.Authenticate()\n",
    "#ee.Initialize()\n",
    "\n",
    "#Map = geemap.Map()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "^C\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting geemap\n",
      "  Using cached geemap-0.20.4-py2.py3-none-any.whl (2.1 MB)\n",
      "Collecting colour\n",
      "  Using cached colour-0.1.5-py2.py3-none-any.whl (23 kB)\n",
      "Requirement already satisfied: pillow in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from geemap) (8.3.2)\n",
      "Collecting bqplot\n",
      "  Using cached bqplot-0.12.42-py2.py3-none-any.whl (1.2 MB)\n",
      "Collecting ipytree\n",
      "  Using cached ipytree-0.2.2-py2.py3-none-any.whl (1.3 MB)\n",
      "Collecting xyzservices\n",
      "  Using cached xyzservices-2023.2.0-py3-none-any.whl (55 kB)\n",
      "Requirement already satisfied: pyshp>=2.1.3 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from geemap) (2.1.3)\n",
      "Collecting whiteboxgui>=0.6.0\n",
      "  Using cached whiteboxgui-2.3.0-py2.py3-none-any.whl (108 kB)\n",
      "Collecting folium>=0.11.0\n",
      "  Using cached folium-0.15.1-py2.py3-none-any.whl (97 kB)\n",
      "Collecting sankee>=0.1.0\n",
      "  Using cached sankee-0.2.4-py3-none-any.whl\n",
      "Collecting pycrs\n",
      "  Using cached PyCRS-1.0.2-py3-none-any.whl\n",
      "Requirement already satisfied: pandas in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from geemap) (1.2.3)\n",
      "Collecting ipyevents\n",
      "  Using cached ipyevents-2.0.2-py3-none-any.whl (101 kB)\n",
      "Collecting ipyleaflet>=0.17.0\n",
      "  Using cached ipyleaflet-0.18.1-py3-none-any.whl (3.7 MB)\n",
      "Collecting earthengine-api>=0.1.304\n",
      "  Using cached earthengine-api-0.1.384.tar.gz (261 kB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "    Preparing wheel metadata: started\n",
      "    Preparing wheel metadata: finished with status 'done'\n",
      "  Using cached earthengine-api-0.1.383.tar.gz (261 kB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "    Preparing wheel metadata: started\n",
      "    Preparing wheel metadata: finished with status 'done'\n",
      "  Using cached earthengine-api-0.1.382.tar.gz (261 kB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "    Preparing wheel metadata: started\n",
      "    Preparing wheel metadata: finished with status 'done'\n",
      "  Using cached earthengine-api-0.1.381.tar.gz (260 kB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "    Preparing wheel metadata: started\n",
      "    Preparing wheel metadata: finished with status 'done'\n",
      "  Using cached earthengine_api-0.1.379-py3-none-any.whl\n",
      "Requirement already satisfied: ipywidgets<8.0.0 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from geemap) (7.4.2)\n",
      "Collecting eerepr>=0.0.4\n",
      "  Using cached eerepr-0.0.4-py3-none-any.whl (9.7 kB)\n",
      "Collecting ffmpeg-python\n",
      "  Using cached ffmpeg_python-0.2.0-py3-none-any.whl (25 kB)\n",
      "Collecting ee-extra>=0.0.10\n",
      "  Using cached ee_extra-0.0.15-py3-none-any.whl\n",
      "Requirement already satisfied: numpy in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from geemap) (1.20.1)\n",
      "Collecting pyperclip\n",
      "  Using cached pyperclip-1.8.2-py3-none-any.whl\n",
      "Collecting gdown\n",
      "  Using cached gdown-4.7.1-py3-none-any.whl (15 kB)\n",
      "Collecting geocoder\n",
      "  Using cached geocoder-1.38.1-py2.py3-none-any.whl (98 kB)\n",
      "Collecting geeadd>=0.5.1\n",
      "  Using cached geeadd-1.0.1-py3-none-any.whl (27 kB)\n",
      "Collecting ipyfilechooser>=0.6.0\n",
      "  Using cached ipyfilechooser-0.6.0-py3-none-any.whl (11 kB)\n",
      "Collecting python-box\n",
      "  Using cached python_box-7.0.1-cp37-cp37m-win_amd64.whl (1.0 MB)\n",
      "Requirement already satisfied: matplotlib in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from geemap) (3.4.2)\n",
      "Collecting scooby\n",
      "  Using cached scooby-0.7.3-py3-none-any.whl (16 kB)\n",
      "Collecting geojson\n",
      "  Using cached geojson-3.1.0-py3-none-any.whl (15 kB)\n",
      "Collecting httplib2<1dev,>=0.9.2\n",
      "  Using cached httplib2-0.22.0-py3-none-any.whl (96 kB)\n",
      "Collecting google-auth-httplib2>=0.0.3\n",
      "  Using cached google_auth_httplib2-0.2.0-py2.py3-none-any.whl (9.3 kB)\n",
      "Requirement already satisfied: requests in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from earthengine-api>=0.1.304->geemap) (2.25.1)\n",
      "Collecting google-api-python-client>=1.12.1\n",
      "  Using cached google_api_python_client-2.111.0-py2.py3-none-any.whl (13.0 MB)\n",
      "Requirement already satisfied: google-auth>=1.4.1 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from earthengine-api>=0.1.304->geemap) (2.6.0)\n",
      "Requirement already satisfied: google-cloud-storage in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from earthengine-api>=0.1.304->geemap) (2.0.0)\n",
      "Collecting branca>=0.6.0\n",
      "  Using cached branca-0.7.0-py3-none-any.whl (25 kB)\n",
      "Requirement already satisfied: jinja2>=2.9 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from folium>=0.11.0->geemap) (2.11.3)\n",
      "Collecting logzero>=1.5.0\n",
      "  Using cached logzero-1.7.0-py2.py3-none-any.whl (16 kB)\n",
      "Collecting beautifulsoup4>=4.9.0\n",
      "  Using cached beautifulsoup4-4.12.2-py3-none-any.whl (142 kB)\n",
      "Collecting soupsieve>1.2\n",
      "  Using cached soupsieve-2.4.1-py3-none-any.whl (36 kB)\n",
      "Requirement already satisfied: google-api-core!=2.0.*,!=2.1.*,!=2.2.*,!=2.3.0,<3.0.0.dev0,>=1.31.5 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from google-api-python-client>=1.12.1->earthengine-api>=0.1.304->geemap) (2.4.0)\n",
      "Collecting uritemplate<5,>=3.0.1\n",
      "  Using cached uritemplate-4.1.1-py2.py3-none-any.whl (10 kB)\n",
      "Requirement already satisfied: protobuf>=3.12.0 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from google-api-core!=2.0.*,!=2.1.*,!=2.2.*,!=2.3.0,<3.0.0.dev0,>=1.31.5->google-api-python-client>=1.12.1->earthengine-api>=0.1.304->geemap) (3.19.4)\n",
      "Requirement already satisfied: googleapis-common-protos<2.0dev,>=1.52.0 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from google-api-core!=2.0.*,!=2.1.*,!=2.2.*,!=2.3.0,<3.0.0.dev0,>=1.31.5->google-api-python-client>=1.12.1->earthengine-api>=0.1.304->geemap) (1.54.0)\n",
      "Requirement already satisfied: setuptools>=40.3.0 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from google-api-core!=2.0.*,!=2.1.*,!=2.2.*,!=2.3.0,<3.0.0.dev0,>=1.31.5->google-api-python-client>=1.12.1->earthengine-api>=0.1.304->geemap) (52.0.0.post20210125)\n",
      "Requirement already satisfied: pyasn1-modules>=0.2.1 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from google-auth>=1.4.1->earthengine-api>=0.1.304->geemap) (0.2.8)\n",
      "Requirement already satisfied: cachetools<6.0,>=2.0.0 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from google-auth>=1.4.1->earthengine-api>=0.1.304->geemap) (4.2.4)\n",
      "Requirement already satisfied: rsa<5,>=3.1.4 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from google-auth>=1.4.1->earthengine-api>=0.1.304->geemap) (4.8)\n",
      "Requirement already satisfied: six>=1.9.0 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from google-auth>=1.4.1->earthengine-api>=0.1.304->geemap) (1.15.0)\n",
      "Requirement already satisfied: pyparsing!=3.0.0,!=3.0.1,!=3.0.2,!=3.0.3,<4,>=2.4.2 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from httplib2<1dev,>=0.9.2->earthengine-api>=0.1.304->geemap) (2.4.7)\n",
      "Collecting traittypes<3,>=0.2.1\n",
      "  Using cached traittypes-0.2.1-py2.py3-none-any.whl (8.6 kB)\n",
      "Collecting ipywidgets<8.0.0\n",
      "  Using cached ipywidgets-7.8.1-py2.py3-none-any.whl (124 kB)\n",
      "Requirement already satisfied: traitlets>=4.3.1 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from ipywidgets<8.0.0->geemap) (4.3.3)\n",
      "Requirement already satisfied: ipython-genutils~=0.2.0 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from ipywidgets<8.0.0->geemap) (0.2.0)\n",
      "Collecting comm>=0.1.3\n",
      "  Using cached comm-0.1.4-py3-none-any.whl (6.6 kB)\n",
      "Collecting jupyterlab-widgets<3,>=1.0.0\n",
      "  Using cached jupyterlab_widgets-1.1.7-py3-none-any.whl (295 kB)\n",
      "Requirement already satisfied: ipython>=4.0.0 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from ipywidgets<8.0.0->geemap) (7.21.0)\n",
      "Collecting widgetsnbextension~=3.6.6\n",
      "  Using cached widgetsnbextension-3.6.6-py2.py3-none-any.whl (1.6 MB)\n",
      "Requirement already satisfied: decorator in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from ipython>=4.0.0->ipywidgets<8.0.0->geemap) (4.4.2)\n",
      "Requirement already satisfied: jedi>=0.16 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from ipython>=4.0.0->ipywidgets<8.0.0->geemap) (0.18.0)\n",
      "Requirement already satisfied: pygments in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from ipython>=4.0.0->ipywidgets<8.0.0->geemap) (2.9.0)\n",
      "Requirement already satisfied: colorama in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from ipython>=4.0.0->ipywidgets<8.0.0->geemap) (0.4.4)\n",
      "Requirement already satisfied: pickleshare in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from ipython>=4.0.0->ipywidgets<8.0.0->geemap) (0.7.5)\n",
      "Requirement already satisfied: prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from ipython>=4.0.0->ipywidgets<8.0.0->geemap) (3.0.5)\n",
      "Requirement already satisfied: backcall in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from ipython>=4.0.0->ipywidgets<8.0.0->geemap) (0.2.0)\n",
      "Requirement already satisfied: parso<0.9.0,>=0.8.0 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from jedi>=0.16->ipython>=4.0.0->ipywidgets<8.0.0->geemap) (0.8.1)\n",
      "Requirement already satisfied: MarkupSafe>=0.23 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from jinja2>=2.9->folium>=0.11.0->geemap) (1.1.1)\n",
      "Requirement already satisfied: wcwidth in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0->ipython>=4.0.0->ipywidgets<8.0.0->geemap) (0.2.5)\n",
      "Requirement already satisfied: pyasn1<0.5.0,>=0.4.6 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from pyasn1-modules>=0.2.1->google-auth>=1.4.1->earthengine-api>=0.1.304->geemap) (0.4.8)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from requests->earthengine-api>=0.1.304->geemap) (1.26.6)\n",
      "Requirement already satisfied: idna<3,>=2.5 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from requests->earthengine-api>=0.1.304->geemap) (2.10)\n",
      "Requirement already satisfied: chardet<5,>=3.0.2 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from requests->earthengine-api>=0.1.304->geemap) (4.0.0)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from requests->earthengine-api>=0.1.304->geemap) (2021.5.30)\n",
      "Collecting plotly>=5.2.2\n",
      "  Using cached plotly-5.18.0-py3-none-any.whl (15.6 MB)\n",
      "Collecting tenacity>=6.2.0\n",
      "  Using cached tenacity-8.2.3-py3-none-any.whl (24 kB)\n",
      "Requirement already satisfied: packaging in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from plotly>=5.2.2->sankee>=0.1.0->geemap) (20.9)\n",
      "Collecting whitebox\n",
      "  Using cached whitebox-2.3.1-py2.py3-none-any.whl (72 kB)\n",
      "Requirement already satisfied: notebook>=4.4.1 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from widgetsnbextension~=3.6.6->ipywidgets<8.0.0->geemap) (5.7.10)\n",
      "Requirement already satisfied: ipykernel in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.6->ipywidgets<8.0.0->geemap) (5.1.1)\n",
      "Requirement already satisfied: terminado>=0.8.1 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.6->ipywidgets<8.0.0->geemap) (0.9.3)\n",
      "Requirement already satisfied: Send2Trash in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.6->ipywidgets<8.0.0->geemap) (1.5.0)\n",
      "Requirement already satisfied: nbconvert<6.0 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.6->ipywidgets<8.0.0->geemap) (5.6.1)\n",
      "Requirement already satisfied: prometheus-client in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.6->ipywidgets<8.0.0->geemap) (0.8.0)\n",
      "Requirement already satisfied: tornado<7,>=4.1 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.6->ipywidgets<8.0.0->geemap) (6.1)\n",
      "Requirement already satisfied: jupyter-core>=4.4.0 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.6->ipywidgets<8.0.0->geemap) (4.6.3)\n",
      "Requirement already satisfied: pyzmq>=17 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.6->ipywidgets<8.0.0->geemap) (22.2.1)\n",
      "Requirement already satisfied: nbformat in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.6->ipywidgets<8.0.0->geemap) (5.1.3)\n",
      "Requirement already satisfied: jupyter-client>=5.2.0 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from notebook>=4.4.1->widgetsnbextension~=3.6.6->ipywidgets<8.0.0->geemap) (6.1.7)\n",
      "Requirement already satisfied: python-dateutil>=2.1 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from jupyter-client>=5.2.0->notebook>=4.4.1->widgetsnbextension~=3.6.6->ipywidgets<8.0.0->geemap) (2.8.1)\n",
      "Requirement already satisfied: pandocfilters>=1.4.1 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from nbconvert<6.0->notebook>=4.4.1->widgetsnbextension~=3.6.6->ipywidgets<8.0.0->geemap) (1.4.3)\n",
      "Requirement already satisfied: defusedxml in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from nbconvert<6.0->notebook>=4.4.1->widgetsnbextension~=3.6.6->ipywidgets<8.0.0->geemap) (0.7.1)\n",
      "Requirement already satisfied: testpath in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from nbconvert<6.0->notebook>=4.4.1->widgetsnbextension~=3.6.6->ipywidgets<8.0.0->geemap) (0.4.4)\n",
      "Requirement already satisfied: bleach in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from nbconvert<6.0->notebook>=4.4.1->widgetsnbextension~=3.6.6->ipywidgets<8.0.0->geemap) (3.3.0)\n",
      "Requirement already satisfied: mistune<2,>=0.8.1 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from nbconvert<6.0->notebook>=4.4.1->widgetsnbextension~=3.6.6->ipywidgets<8.0.0->geemap) (0.8.4)\n",
      "Requirement already satisfied: entrypoints>=0.2.2 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from nbconvert<6.0->notebook>=4.4.1->widgetsnbextension~=3.6.6->ipywidgets<8.0.0->geemap) (0.3)\n",
      "Requirement already satisfied: jsonschema!=2.5.0,>=2.4 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from nbformat->notebook>=4.4.1->widgetsnbextension~=3.6.6->ipywidgets<8.0.0->geemap) (3.2.0)\n",
      "Requirement already satisfied: attrs>=17.4.0 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from jsonschema!=2.5.0,>=2.4->nbformat->notebook>=4.4.1->widgetsnbextension~=3.6.6->ipywidgets<8.0.0->geemap) (21.2.0)\n",
      "Requirement already satisfied: pyrsistent>=0.14.0 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from jsonschema!=2.5.0,>=2.4->nbformat->notebook>=4.4.1->widgetsnbextension~=3.6.6->ipywidgets<8.0.0->geemap) (0.17.3)\n",
      "Requirement already satisfied: importlib-metadata in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from jsonschema!=2.5.0,>=2.4->nbformat->notebook>=4.4.1->widgetsnbextension~=3.6.6->ipywidgets<8.0.0->geemap) (3.7.3)\n",
      "Requirement already satisfied: pywinpty>=0.5 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from terminado>=0.8.1->notebook>=4.4.1->widgetsnbextension~=3.6.6->ipywidgets<8.0.0->geemap) (0.5.7)\n",
      "Requirement already satisfied: webencodings in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from bleach->nbconvert<6.0->notebook>=4.4.1->widgetsnbextension~=3.6.6->ipywidgets<8.0.0->geemap) (0.5.1)\n",
      "Requirement already satisfied: pytz>=2017.3 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages\\pytz-2020.1-py3.7.egg (from pandas->geemap) (2020.1)\n",
      "Requirement already satisfied: future in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages\\future-0.18.2-py3.7.egg (from ffmpeg-python->geemap) (0.18.2)\n",
      "Collecting filelock\n",
      "  Using cached filelock-3.12.2-py3-none-any.whl (10 kB)\n",
      "Collecting tqdm\n",
      "  Using cached tqdm-4.66.1-py3-none-any.whl (78 kB)\n",
      "Requirement already satisfied: click in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from geocoder->geemap) (7.1.2)\n",
      "Collecting ratelim\n",
      "  Using cached ratelim-0.1.6-py2.py3-none-any.whl (4.0 kB)\n",
      "Requirement already satisfied: google-cloud-core<3.0dev,>=1.6.0 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from google-cloud-storage->earthengine-api>=0.1.304->geemap) (2.2.2)\n",
      "Requirement already satisfied: google-resumable-media>=1.3.0 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from google-cloud-storage->earthengine-api>=0.1.304->geemap) (2.1.0)\n",
      "Requirement already satisfied: google-crc32c<2.0dev,>=1.0 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from google-resumable-media>=1.3.0->google-cloud-storage->earthengine-api>=0.1.304->geemap) (1.3.0)\n",
      "Requirement already satisfied: typing-extensions>=3.6.4 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from importlib-metadata->jsonschema!=2.5.0,>=2.4->nbformat->notebook>=4.4.1->widgetsnbextension~=3.6.6->ipywidgets<8.0.0->geemap) (3.7.4.3)\n",
      "Requirement already satisfied: zipp>=0.5 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from importlib-metadata->jsonschema!=2.5.0,>=2.4->nbformat->notebook>=4.4.1->widgetsnbextension~=3.6.6->ipywidgets<8.0.0->geemap) (3.4.1)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from matplotlib->geemap) (1.3.1)\n",
      "Requirement already satisfied: cycler>=0.10 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from matplotlib->geemap) (0.10.0)\n",
      "Requirement already satisfied: PySocks!=1.5.7,>=1.5.6 in c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages (from requests->earthengine-api>=0.1.304->geemap) (1.7.1)\n",
      "Installing collected packages: httplib2, widgetsnbextension, uritemplate, jupyterlab-widgets, google-auth-httplib2, comm, tenacity, soupsieve, ipywidgets, google-api-python-client, xyzservices, whitebox, traittypes, tqdm, ratelim, plotly, logzero, ipytree, ipyfilechooser, filelock, earthengine-api, branca, beautifulsoup4, whiteboxgui, scooby, sankee, python-box, pyperclip, pycrs, ipyleaflet, ipyevents, geojson, geocoder, geeadd, gdown, folium, ffmpeg-python, eerepr, ee-extra, colour, bqplot, geemap\n",
      "  Attempting uninstall: widgetsnbextension\n",
      "    Found existing installation: widgetsnbextension 3.4.2\n",
      "    Uninstalling widgetsnbextension-3.4.2:\n",
      "      Successfully uninstalled widgetsnbextension-3.4.2\n",
      "  Attempting uninstall: ipywidgets\n",
      "    Found existing installation: ipywidgets 7.4.2\n",
      "    Uninstalling ipywidgets-7.4.2:\n",
      "      Successfully uninstalled ipywidgets-7.4.2\n",
      "Successfully installed beautifulsoup4-4.12.2 bqplot-0.12.42 branca-0.7.0 colour-0.1.5 comm-0.1.4 earthengine-api-0.1.379 ee-extra-0.0.15 eerepr-0.0.4 ffmpeg-python-0.2.0 filelock-3.12.2 folium-0.15.1 gdown-4.7.1 geeadd-1.0.1 geemap-0.20.4 geocoder-1.38.1 geojson-3.1.0 google-api-python-client-2.111.0 google-auth-httplib2-0.2.0 httplib2-0.22.0 ipyevents-2.0.2 ipyfilechooser-0.6.0 ipyleaflet-0.18.1 ipytree-0.2.2 ipywidgets-7.8.1 jupyterlab-widgets-1.1.7 logzero-1.7.0 plotly-5.18.0 pycrs-1.0.2 pyperclip-1.8.2 python-box-7.0.1 ratelim-0.1.6 sankee-0.2.4 scooby-0.7.3 soupsieve-2.4.1 tenacity-8.2.3 tqdm-4.66.1 traittypes-0.2.1 uritemplate-4.1.1 whitebox-2.3.1 whiteboxgui-2.3.0 widgetsnbextension-3.6.6 xyzservices-2023.2.0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "  WARNING: Generating metadata for package earthengine-api produced metadata for project name unknown. Fix your #egg=earthengine-api fragments.\n",
      "WARNING: Discarding https://files.pythonhosted.org/packages/96/62/b7651eeedacbbaae1840f1ae902269160bd40ea4861fc99ca0d4eab57115/earthengine-api-0.1.384.tar.gz#sha256=b2ed22d5fa9a930f6b97c439b99e51fea75f4d8e911ead740d3bce540dabdf6b (from https://pypi.org/simple/earthengine-api/) (requires-python:>=3.7). Requested unknown from https://files.pythonhosted.org/packages/96/62/b7651eeedacbbaae1840f1ae902269160bd40ea4861fc99ca0d4eab57115/earthengine-api-0.1.384.tar.gz#sha256=b2ed22d5fa9a930f6b97c439b99e51fea75f4d8e911ead740d3bce540dabdf6b (from geemap) has inconsistent name: filename has 'earthengine-api', but metadata has 'UNKNOWN'\n",
      "  WARNING: Generating metadata for package earthengine-api produced metadata for project name unknown. Fix your #egg=earthengine-api fragments.\n",
      "WARNING: Discarding https://files.pythonhosted.org/packages/31/ee/57e478381f386a4a9b7d3a9fc2df20bb835a0eed9614d66be5078bffa094/earthengine-api-0.1.383.tar.gz#sha256=3e90b493f59b87fc3ea358d5aba9828ee280d36e1bb96a2ea0d899aaf741ff41 (from https://pypi.org/simple/earthengine-api/) (requires-python:>=3.7). Requested unknown from https://files.pythonhosted.org/packages/31/ee/57e478381f386a4a9b7d3a9fc2df20bb835a0eed9614d66be5078bffa094/earthengine-api-0.1.383.tar.gz#sha256=3e90b493f59b87fc3ea358d5aba9828ee280d36e1bb96a2ea0d899aaf741ff41 (from geemap) has inconsistent name: filename has 'earthengine-api', but metadata has 'UNKNOWN'\n",
      "  WARNING: Generating metadata for package earthengine-api produced metadata for project name unknown. Fix your #egg=earthengine-api fragments.\n",
      "WARNING: Discarding https://files.pythonhosted.org/packages/3f/63/530b16ca58c7e94d583fc2db32cc9131efaedb917399710a02e80cf273da/earthengine-api-0.1.382.tar.gz#sha256=5cc35adcd24616f2c31f9e4ae1001f4efdaf4da7f41d4de51bf831c711c97cbf (from https://pypi.org/simple/earthengine-api/) (requires-python:>=3.7). Requested unknown from https://files.pythonhosted.org/packages/3f/63/530b16ca58c7e94d583fc2db32cc9131efaedb917399710a02e80cf273da/earthengine-api-0.1.382.tar.gz#sha256=5cc35adcd24616f2c31f9e4ae1001f4efdaf4da7f41d4de51bf831c711c97cbf (from geemap) has inconsistent name: filename has 'earthengine-api', but metadata has 'UNKNOWN'\n",
      "  WARNING: Generating metadata for package earthengine-api produced metadata for project name unknown. Fix your #egg=earthengine-api fragments.\n",
      "WARNING: Discarding https://files.pythonhosted.org/packages/2b/b5/ebe8dd2571bea3ab5a5782dc575c5fdafcf53450e06dc0fbf4383fc332c3/earthengine-api-0.1.381.tar.gz#sha256=34e9d53018fe30ac3f7c33620edf08e22058619be06748950f9ca48d82678b9f (from https://pypi.org/simple/earthengine-api/) (requires-python:>=3.7). Requested unknown from https://files.pythonhosted.org/packages/2b/b5/ebe8dd2571bea3ab5a5782dc575c5fdafcf53450e06dc0fbf4383fc332c3/earthengine-api-0.1.381.tar.gz#sha256=34e9d53018fe30ac3f7c33620edf08e22058619be06748950f9ca48d82678b9f (from geemap) has inconsistent name: filename has 'earthengine-api', but metadata has 'UNKNOWN'\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "    WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "    WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\users\\raul.perez.astorga\\appdata\\local\\esri\\conda\\envs\\arcgispro-py3-clone-1\\lib\\site-packages)\n"
     ]
    }
   ],
   "source": [
    "%pip install geemap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "t1 = '2023-12-16'\n",
    "t2 = '2024-02-24'\n",
    "roi = ee.FeatureCollection('users/raulperezastorga/Limites_SPPP')\n",
    "\n",
    "def addNBR_s2h(image):\n",
    "    nbr = image.normalizedDifference(['B8', 'B12']).rename(\"nbr\")\n",
    "    image = image.addBands(nbr)\n",
    "    return image\n",
    "\n",
    "def mask_fire_firms(image):\n",
    "    firms = ee.ImageCollection('FIRMS')\n",
    "    fires_firms = ee.ImageCollection(firms\\\n",
    "            .select('T21')\\\n",
    "            .filterDate(t1,t2)\\\n",
    "            .filterBounds(roi)) \n",
    "    fires_firmsmx = fires_firms.max().clip(roi).subtract(273.15)\n",
    "    fires_firmsrc = fires_firmsmx.gte(100)\n",
    "    return image.updateMask(fires_firmsrc)\n",
    "\n",
    "# Define a function to apply the vegetation mask to an input image\n",
    "def maskVegetation(image):\n",
    "    mcd12q1 = ee.ImageCollection('MODIS/006/MCD12Q1') \\\n",
    "        .filter(ee.Filter.date(t1, t2)) \\\n",
    "        .filterBounds(roi)\\\n",
    "        .first()\n",
    "    vg = mcd12q1.select('LC_Type1')\n",
    "    # Get the pixels that represent vegetation\n",
    "    mask = vg.eq(1).Or(vg.eq(2)).Or(vg.eq(3)).Or(vg.eq(4)).Or(vg.eq(5)).Or(vg.eq(6)).Or(vg.eq(7))\n",
    "    \n",
    "    # Return the masked image with updated metadata\n",
    "    return image.updateMask(mask)\n",
    "\n",
    "# mask elevation\n",
    "def mask_elevation(image):\n",
    "    elevation_data = ee.Image('USGS/SRTMGL1_003').select('elevation').clip(roi)\n",
    "    elevation_mask = elevation_data.lt(3500)\n",
    "    return image.updateMask(elevation_mask)\n",
    "\n",
    "\n",
    "firms = ee.ImageCollection('FIRMS')\n",
    "fires_firms = ee.ImageCollection(firms\\\n",
    "        .select('T21')\\\n",
    "        .filterDate(t1,t2)\\\n",
    "        .filterBounds(roi)) \n",
    "fires_firmsmx = fires_firms.max().clip(roi).subtract(273.15)\n",
    "fires_firmsrc = fires_firmsmx.gte(100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "s2 = ee.ImageCollection('COPERNICUS/S2_HARMONIZED') \\\n",
    "    .filterDate(t1, t2) \\\n",
    "    .filter(ee.Filter.lt('CLOUDY_PIXEL_PERCENTAGE', 40)) \\\n",
    "    .filterBounds(roi) \\\n",
    "    .sort('system:time_start', True) \\\n",
    "    .map(lambda image: image.addBands(image.expression('(RED * 1.53)', { 'RED': image.select('B4') }).rename('B4B')))\n",
    "\n",
    "vis = {\n",
    "    'bands': ['B8', 'B4B', 'B3'],\n",
    "    'min': 0,\n",
    "    'max': 3800,\n",
    "    'gamma': [1, 1, 1]\n",
    "}\n",
    "\n",
    "img = s2.mosaic().clip(roi)\n",
    "# Use geemap to display the image\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Map.addLayer(img, vis, 'falso')\n",
    "Map.layers"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
